command/agentproxyshared/cache/static_secret_capability_manager.go

```diff
@@ -0,0 +1,260 @@
+				index.IndexLock.Lock()
+				delete(index.Tokens, capabilitiesIndex.Token)
+				err = sscm.leaseCache.db.Set(index)
+				if err != nil {
+					sscm.logger.Error("error when attempting to update index in cache", "index.ID", index.ID, "err", err)
+				}
+				index.IndexLock.Unlock()
+			}
+		}
+
+		// Lastly, we should update the capabilities index, either evicting or updating it
+		capabilitiesIndex.IndexLock.Lock()
+		defer capabilitiesIndex.IndexLock.Unlock()
+		if len(newReadablePaths) == 0 {
+			err := sscm.leaseCache.db.EvictCapabilitiesIndex(cachememdb.IndexNameID, indexToRenew.ID)
+			if err != nil {
+				sscm.logger.Error("error when attempting to evict capabilities from cache", "index.ID", indexToRenew.ID, "err", err)
+				sscm.submitWorkToPoolAfterInterval(work)
+				return
+			}
+			// If we successfully evicted the index, no need to re-submit the work to the pool.
+			return
+		}
+
+		// The token still has some capabilities, so, update the capabilities index:
+		capabilitiesIndex.ReadablePaths = newReadablePaths
+		err = sscm.leaseCache.db.SetCapabilitiesIndex(capabilitiesIndex)
+		if err != nil {
+			sscm.logger.Error("error when attempting to update capabilities from cache", "index.ID", indexToRenew.ID, "err", err)
+		}
+
+		// Finally, put ourselves back on the work pool after
+		sscm.submitWorkToPoolAfterInterval(work)
+		return
+	}
+
+	sscm.submitWorkToPoolAfterInterval(work)
+}
+
+// getCapabilities is a wrapper around a /sys/capabilities-self call that returns
+// capabilities as a map with paths as keys, and capabilities as values.
+func getCapabilities(paths []string, client *api.Client) (map[string][]string, error) {
+	body := make(map[string]interface{})
+	body["paths"] = paths
+	capabilities := make(map[string][]string)
+
+	secret, err := client.Logical().Write("sys/capabilities-self", body)
+	if err != nil && strings.Contains(err.Error(), "permission denied") {
+		// Token has expired. Return an empty set of capabilities:
+		return capabilities, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if secret == nil || secret.Data == nil {
+		return nil, errors.New("data from server response is empty")
+	}
+
+	for _, path := range paths {
+		var res []string
+		err = mapstructure.Decode(secret.Data[path], &res)
+		if err != nil {
+			return nil, err
+		}
+
+		capabilities[path] = res
+	}
+
+	return capabilities, nil
+}
+
+// reconcileCapabilities takes a set of known readable paths, and a set of capabilities (a response from the
+// sys/capabilities-self endpoint) and returns a subset of the readablePaths after taking into account any updated
+// capabilities as a set, represented by a map of strings to structs.
+// It will delete any path in readablePaths if it does not have a "root" or "read" capability listed in the
+// capabilities map.
+func reconcileCapabilities(readablePaths []string, capabilities map[string][]string) map[string]struct{} {
```

Comment:
I ensured this was its own function for testing purposes - I tried to ensure this was well-tested.